README.md
# Spam Assassin

SpamAssassin is a tool to classify spam messages or not using classifer technique in machine learning. 
You also can add more data to train this machine learning become better to classify spam messages or not by re-evaluate the test data, and save it to train data.

## Getting Started

These application is requirements for Laboratory of Graphics and Artificial Intelligence.

Feature :
- Evaluate and predict the test data
- Open arff file (using Weka Library)
- GUI

## Screenshot
![MainWindow](https://raw.githubusercontent.com/ironlota/spamassassin/master/image/mainwindow.png)
![Tested](https://raw.githubusercontent.com/ironlota/spamassassin/master/image/tested.png)
![Classifier](https://raw.githubusercontent.com/ironlota/spamassassin/master/image/classifier.png)
![Train](https://raw.githubusercontent.com/ironlota/spamassassin/master/image/train.png)
![ChooseFile](https://raw.githubusercontent.com/ironlota/spamassassin/master/image/choosefile.png)

### Prerequisites

What things you need to install the software and how to install them

```
Installed :
- Java Development Kit (version > 7.0)
- Java IDE (Intellij, Netbeans, or Eclipse)
- Gradle
```

What things you need to include in project :

```
Libraries :
- Weka (www.cs.waikato.ac.nz/ml/weka/) -> [library for Machine Learning]
```

### Installing

```
wget [https://github.com/ironlota/spamassassin/archive/master.zip](https://github.com/ironlota/spamassassin/archive/master.zip)
unzip master.zip
cd spamassassin-master
```

### Running
```
- gradle run (if you do not want to make a jar)
- gradle jfxJar (if you want to make a jar)

Note :
Due to path issue, you have to make sure that you run the jar file on the project root.
This can be happened because "./dataset" folder remains in root folder of project.
```


## Change Log
v.0.1 initial release
- using j48 unpruned tree
- using Indonesian training data set
- using JavaFX for GUI
- using Weka as library

## To do and Bug
- make arff converter from txt file
- show tree in GUI

## Credits
- [Weka 3 - Data Mining Open Source Machine Learning Software](www.cs.waikato.ac.nz/ml/weka/)
- [Machine Learning Recipe with Josh Gordon](https://www.youtube.com/playlist?list=PLOU2XLYxmsIIuiBfYad6rFYQU_jL2ryal)
- [Weka Text Classification for First Time & Beginner Users by Brandon Weinberg](https://www.youtube.com/watch?v=IY29uC4uem8)
- [Text Classification with Weka using a J48 Decision Tree by S0naris](https://www.youtube.com/watch?v=HrixTPMOCD4)
- [JFoenix](www.jfoenix.com)
- [JavaFX Gradle Plugin](https://github.com/FibreFoX/javafx-gradle-plugin)

## Version

0.1 Initial Release

## Authors

* **Ray Andrew** - [Ironlota](https://github.com/ironlota)

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

changelog.txt
v.0.1 initial release
- using j48 unpruned tree
- using Indonesian training data set
- using JavaFX for GUI
- using Weka as library

To do :
- make arff converter from txt file
- show tree in GUI

src/main/java/com/rayandrew/spamassassin/MainClass.java
package com.rayandrew.spamassassin;

import javafx.application.Application;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.stage.Stage;

/**
 * Created by rufus on 6/7/2017.
 */
public class MainClass extends Application {
    public static Stage primaryStage;

    public static void main(String[] args) {
        launch(args);
    }

    @Override
    public void start(Stage stage) throws Exception {
        FXMLLoader loader = new FXMLLoader(getClass().getResource("/MainClass.fxml"));

        Parent root = loader.load();
        InterfaceClass interfaceClass = loader.getController();
        interfaceClass.setHostServices(getHostServices());
        Scene scene = new Scene(root);
        primaryStage = stage;
        scene.getStylesheets().add(getClass().getResource("/modena_dark.css").toExternalForm());
        primaryStage.setTitle("Spam Assassin");
        primaryStage.setScene(scene);
        primaryStage.setResizable(false);
        primaryStage.show();
    }
}

src/main/java/com/rayandrew/spamassassin/SpamResult.java
package com.rayandrew.spamassassin;

import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleStringProperty;

/**
 * Created by ironlota on 6/15/17.
 */
public class SpamResult {
    private final SimpleIntegerProperty no;
    private final SimpleStringProperty text;
    private final SimpleStringProperty result;

    public SpamResult(int no, String text, String result) {
        this.no = new SimpleIntegerProperty(no);
        this.text = new SimpleStringProperty(text);
        this.result = new SimpleStringProperty(result);
    }

    public int getNo() {
        return no.get();
    }

    public void setNo(int no) {
        this.no.set(no);
    }

    public String getText() {
        return text.get();
    }

    public void setText(String text) {
        this.text.set(text);
    }

    public String getResult() {
        return result.get();
    }

    public void setResult(String result) {
        this.result.set(result);
    }
}


src/main/java/com/rayandrew/spamassassin/TrainingClass.java
package com.rayandrew.spamassassin;

import weka.classifiers.meta.FilteredClassifier;
import weka.classifiers.trees.J48;
import weka.core.Instances;
import weka.core.converters.ConverterUtils;
import weka.core.stemmers.IteratedLovinsStemmer;
import weka.core.stopwords.RegExpFromFile;
import weka.core.tokenizers.WordTokenizer;
import weka.filters.unsupervised.attribute.StringToWordVector;

/**
 * Created by ironlota on 6/15/17.
 */
public class TrainingClass {
    public static Instances readDataFile(String filename) throws Exception {
        ConverterUtils.DataSource inputReader = null;
        try {
            inputReader = new ConverterUtils.DataSource(filename);
        } catch (Exception e) {
            System.err.println("File not found: " + filename);
            e.printStackTrace();
        }

        if (inputReader == null) {
            return null;
        } else {
            return inputReader.getDataSet();
        }
    }

    public static StringToWordVector defaultStringToWordVector() {
        StringToWordVector stw = new StringToWordVector();
        stw.setIDFTransform(true);
        stw.setTFTransform(true);
        stw.setAttributeIndices("first-last");
        stw.setDebug(false);
        stw.setPeriodicPruning(-1.0);
        stw.setStemmer(new IteratedLovinsStemmer());
        stw.setStopwordsHandler(new RegExpFromFile());
        stw.setTokenizer(new WordTokenizer());
        stw.setWordsToKeep(1000);
        stw.setMinTermFreq(1);
        stw.setOutputWordCounts(true);
        return stw;
    }

    public static J48 defaultJ48() {
        J48 j48 = new J48();
        j48.setUnpruned(true);
        j48.setCollapseTree(true);
        j48.setSubtreeRaising(true);
        j48.setBatchSize("100");
        j48.setUseMDLcorrection(true);
        return j48;
    }

    public static FilteredClassifier trainData(Instances train) throws Exception {
        FilteredClassifier fc = new FilteredClassifier();
        fc.setFilter(defaultStringToWordVector());
        fc.setClassifier(defaultJ48());
        fc.buildClassifier(train);
        return fc;
    }
}


src/main/java/com/rayandrew/spamassassin/InterfaceClass.java
package com.rayandrew.spamassassin;

import javafx.application.HostServices;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.Hyperlink;
import javafx.scene.control.Label;
import javafx.scene.control.MenuItem;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableColumn.CellEditEvent;
import javafx.scene.control.TablePosition;
import javafx.scene.control.TableView;
import javafx.scene.control.TextArea;
import javafx.scene.control.cell.ComboBoxTableCell;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCodeCombination;
import javafx.scene.input.KeyCombination;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import javafx.scene.text.Text;
import javafx.scene.text.TextFlow;
import javafx.stage.FileChooser;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.util.Callback;
import weka.classifiers.meta.FilteredClassifier;
import weka.core.Instances;
import weka.gui.beans.TextViewer;

import java.awt.Desktop;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.ResourceBundle;
import java.util.Scanner;

import static com.rayandrew.spamassassin.MainClass.primaryStage;

/**
 * Created by ironlota on 6/15/17.
 */
public class InterfaceClass implements Initializable {
    private ObservableList<SpamResult> data = FXCollections.observableArrayList();

    @FXML
    private TableView tableView;

    @FXML
    private TableColumn noRow;

    @FXML
    private TableColumn text;

    @FXML
    private TableColumn<SpamResult, SpamStatus> result;

    @FXML
    private MenuItem openFile;

    //@FXML
    //private MenuItem saveFile;

    @FXML
    private MenuItem closeWindow;

    @FXML
    private MenuItem aboutWindow;

    @FXML
    private MenuItem helpItem;

    @FXML
    private MenuItem changeLog;

    @FXML
    private TextArea trainData;

    @FXML
    private TextArea classifierData;

    @FXML
    private Hyperlink urlLabel;

    private File dataFile = null;
    private static Instances train = null;
    private static Instances test = null;

    private HostServices hostServices;

    public HostServices getHostServices() {
        return hostServices ;
    }

    public void setHostServices(HostServices hostServices) {
        this.hostServices = hostServices ;
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        tableView.setEditable(true);

        /**
         * KEY BINDING
         */
        openFile.setAccelerator(new KeyCodeCombination(KeyCode.O, KeyCombination.SHORTCUT_DOWN));
        //saveFile.setAccelerator(new KeyCodeCombination(KeyCode.S, KeyCombination.SHORTCUT_DOWN));
        closeWindow.setAccelerator(new KeyCodeCombination(KeyCode.F4, KeyCombination.ALT_DOWN));
        helpItem.setAccelerator(new KeyCodeCombination(KeyCode.F11));
        changeLog.setAccelerator(new KeyCodeCombination(KeyCode.F12));
        aboutWindow.setAccelerator(new KeyCodeCombination(KeyCode.F10));


        aboutWindow.setOnAction(event -> {
            Alert alert = new Alert(Alert.AlertType.INFORMATION);
            alert.setResizable(true);
            alert.setTitle("About");
            alert.setHeaderText("SpamAssassin v.0.1-beta");
            alert.setContentText("Created by Ray Andrew");
            Hyperlink hyperlink = new Hyperlink("Fork at https://github.com/ironlota/SpamAssassin");
            hyperlink.setOnAction(eventl -> {
                goToGithub();
            });

            Label label = new Label("What can Spam Assassin do?");

            String explanationText = "Spam Assassin is a tool to classify spam messages " +
                    "or not using classifer technique in machine learning." +
                    "\nYou also can add more data to train this machine learning become better " +
                    "\nto classify spam messages or not by re-evaluate the test data, " +
                    "and save it to train data.";

            String specialCredit = "\n\nCredits :\n" +
                    "- Weka 3 - Data Mining Open Source Machine Learning Software (www.cs.waikato.ac.nz/ml/weka/)\n" +
                    "- Machine Learning Recipe with Josh Gordon (https://www.youtube.com/playlist?list=PLOU2XLYxmsIIuiBfYad6rFYQU_jL2ryal)\n" +
                    "- Weka Text Classification for First Time & Beginner Users by Brandon Weinberg (https://www.youtube.com/watch?v=IY29uC4uem8)\n" +
                    "- Text Classification with Weka using a J48 Decision Tree by S0naris (https://www.youtube.com/watch?v=HrixTPMOCD4)\n" +
                    "- JFoenix (www.jfoenix.com)";

            TextArea textArea = new TextArea(explanationText + specialCredit);
            textArea.setEditable(false);
            textArea.setWrapText(true);

            textArea.setMaxWidth(Double.MAX_VALUE);
            textArea.setMaxHeight(Double.MAX_VALUE);
            GridPane.setVgrow(textArea, Priority.ALWAYS);
            GridPane.setHgrow(textArea, Priority.ALWAYS);

            GridPane expContent = new GridPane();
            expContent.setMaxWidth(Double.MAX_VALUE);
            expContent.add(hyperlink, 0, 0);
            expContent.add(label, 0, 1);
            expContent.add(textArea, 0, 2);

            // Set expandable Exception into the dialog pane.
            alert.getDialogPane().setExpandableContent(expContent);

            alert.showAndWait();
        });

        changeLog.setOnAction(event -> {
            Alert alert = new Alert(Alert.AlertType.INFORMATION);
            alert.setResizable(true);
            alert.setTitle("Change Log");
            alert.setHeaderText("SpamAssassin v.0.1-beta");
            alert.setContentText("This project is still in initial version\n");

            Label label = new Label("Change Log");


            StringBuilder str = new StringBuilder();
            try {
                BufferedReader reader = new BufferedReader(new FileReader("changelog.txt"));
                String line;
                while ((line = reader.readLine()) != null)
                {
                    str.append(line + "\n");
                }
                reader.close();
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }
            TextArea textArea = new TextArea(str.toString());

            textArea.setEditable(false);
            textArea.setWrapText(true);

            textArea.setMaxWidth(Double.MAX_VALUE);
            textArea.setMaxHeight(Double.MAX_VALUE);
            GridPane.setVgrow(textArea, Priority.ALWAYS);
            GridPane.setHgrow(textArea, Priority.ALWAYS);

            GridPane expContent = new GridPane();
            expContent.setMaxWidth(Double.MAX_VALUE);
            expContent.add(label, 0, 0);
            expContent.add(textArea, 0, 1);

            // Set expandable Exception into the dialog pane.
            alert.getDialogPane().setExpandableContent(expContent);

            alert.showAndWait();
        });
    }

    @FXML
    public void openAction(ActionEvent actionEvent) {
        handleOpenClick();
    }

    @FXML
    public void quitAction(ActionEvent actionEvent) {
        System.exit(0);
    }

    private void evaluateData(String filename) {
        try {
            train = TrainingClass.readDataFile("./dataset/train.arff");
            test = TrainingClass.readDataFile(filename);
            train.setClassIndex(train.numAttributes() - 1);
            test.setClassIndex(test.numAttributes() - 1);

            trainData.setText(train.toString());


            FilteredClassifier fc = TrainingClass.trainData(train);

            int count = 0;
            for (int i = 0; i < test.numInstances(); i++) {
                double pred = fc.classifyInstance(test.instance(i));
                String actual = test.classAttribute().value((int) test.instance(i).classValue());
                String predicted = test.classAttribute().value((int) pred);
                data.add(new SpamResult(
                        (int) test.instance(i).value(0),
                        test.get(i).stringValue(test.attribute("text")),
                        predicted)
                );
                /**
                 System.out.print("ID: " + test.instance(i).value(0));
                 System.out.print(", actual: " + actual);
                 System.out.println(", predicted: " + predicted);
                 */
                if (!predicted.equals(actual)) {
                    count++;
                }
            }
            //System.out.println("False prediction count = " + count);
            //int sumData = test.numInstances();
            //double s = (sumData - count) * 100 / sumData;
            //System.out.println("Accuracy = " + String.format("%.05f", s));

            classifierData.setText(fc.toString());

            noRow.setCellValueFactory(new PropertyValueFactory<SpamResult, Integer>("no"));

            text.setCellValueFactory(new PropertyValueFactory<>("text"));

            Callback<TableColumn, TableCell> stringCellFactory =
                    p -> {
                        MyStringTableCell cell = new MyStringTableCell();
                        cell.addEventFilter(MouseEvent.MOUSE_CLICKED, event -> {
                            TableCell c = (TableCell) event.getSource();

                            int index = c.getIndex();
                            Alert alert = new Alert(Alert.AlertType.INFORMATION);
                            alert.setResizable(true);
                            alert.setTitle("Text row " + index);
                            alert.setHeaderText("Detailed Text for row " + index);

                            TextArea textArea = new TextArea(data.get(index).getText());

                            textArea.setEditable(false);
                            textArea.setWrapText(true);

                            textArea.setMaxWidth(Double.MAX_VALUE);
                            textArea.setMaxHeight(Double.MAX_VALUE);

                            GridPane.setVgrow(textArea, Priority.ALWAYS);
                            GridPane.setHgrow(textArea, Priority.ALWAYS);

                            GridPane expContent = new GridPane();
                            expContent.setMaxWidth(Double.MAX_VALUE);
                            expContent.add(textArea, 0, 0);

                            alert.getDialogPane().setExpandableContent(expContent);
                            alert.showAndWait();
                        });
                        return cell;
                    };

            text.setCellFactory(stringCellFactory);

            result.setCellValueFactory(new PropertyValueFactory<>("result"));

            ObservableList<SpamStatus> statusList = FXCollections.observableArrayList(//
                    SpamStatus.values());

            result.setCellFactory(ComboBoxTableCell.forTableColumn(statusList));

            result.setOnEditCommit((CellEditEvent<SpamResult, SpamStatus> event) -> {
                TablePosition<SpamResult, SpamStatus> pos = event.getTablePosition();

                SpamStatus newStatus = event.getNewValue();

                int row = pos.getRow();
                SpamResult spam = event.getTableView().getItems().get(row);

                spam.setResult(newStatus.getCode());
            });

            tableView.setItems(data);

        } catch(Exception e) {
            e.printStackTrace();
        }
    }


    private void handleOpenClick() {

        // A fuller JavaFX app would
        // want to perform a check
        // to make sure the current
        // data has been saved.

        //creating JavaFX file chooser
        FileChooser fc = new FileChooser();
        fc.setTitle("Get Arff");
        fc.getExtensionFilters().addAll(
                new FileChooser.ExtensionFilter(
                        "Arff Files",
                        "*.arff"),
                new FileChooser.ExtensionFilter(
                        "Text Files",
                        "*.txt")
        );

        //showing the file chooser
        File phil = fc.showOpenDialog(primaryStage);

        // checking that a file was
        // chosen by the user
        if (phil != null) {

            // opening a scanner
            try (Scanner scan = new Scanner(phil)) {
                // saving the file for
                // use by the saveMI
                dataFile = phil;
                evaluateData(dataFile.getAbsolutePath());
                // enabling saveMI
                //saveMI.setDisable(false);

            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
        }
    }

    class MyStringTableCell extends TableCell<SpamResult, String> {

        @Override
        public void updateItem(String item, boolean empty) {
            super.updateItem(item, empty);
            setText(empty ? null : getString());
            setGraphic(null);
        }

        private String getString() {
            return getItem() == null ? "" : getItem();
        }
    }

    public void goToGithub() {
        hostServices.showDocument("https://github.com/ironlota/SpamAssassin");
    }
}
